infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/replication/receive/LogReplicationSinkManager.java

```diff
@@ -403,17 +412,39 @@ private void processSnapshotMessage(LogReplicationEntry message) {
                 snapshotWriter.apply(message);
                 break;
             case SNAPSHOT_END:
-                setDataConsistent(false);
-                snapshotWriter.snapshotTransferDone(message);
-                completeSnapshotApply(message);
-                setDataConsistent(true);
+                if (snapshotWriter.getPhase() != StreamsSnapshotWriter.Phase.APPLY_PHASE) {
+                    // Mark Snapshot Sync Transfer as complete and return ACK right away
+                    completeSnapshotTransfer(message);
+                    startSnapshotApplyAsync(message);
+                }
                 break;
             default:
-                log.warn("Message type {} should not be applied as snapshot sync.", message.getMetadata().getMessageMetadataType());
+                log.warn("Message type {} should not be applied during snapshot sync.", message.getMetadata().getMessageMetadataType());
                 break;
         }
     }
 
+    private void startSnapshotApplyAsync(LogReplicationEntry entry) {
+        applyExecutor.submit(() -> startSnapshotApply(entry));
+    }
+
+    private void startSnapshotApply(LogReplicationEntry entry) {
+        log.debug("OnEntry :: Snapshot Sync apply, id={}", entry.getMetadata().getSyncRequestId());
+        setDataConsistent(false);
+        snapshotWriter.startSnapshotSyncApply();
+        completeSnapshotApply(entry);
+        setDataConsistent(true);
+        log.debug("OnExit :: Snapshot Sync apply, id={}", entry.getMetadata().getSyncRequestId());
+    }
+
+    private void completeSnapshotTransfer(LogReplicationEntry message) {
+        topologyConfigId = message.getMetadata().getTopologyConfigId();
```

Comment:
Why we need to update topologyConfigId here?

Reply:
I copy/paste the old code, but I don't really think that is needed, removing it.